src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs

```diff
@@ -243,19 +243,14 @@ protected void AddSecrets(TransactionBuildContext context, IEnumerable<ICoin> co
             ExtKey seedExtKey = this.walletManager.GetExtKey(context.AccountReference, context.WalletPassword, context.CacheSecret);
 
             var signingKeys = new HashSet<ISecret>();
-            var added = new HashSet<HdAddress>();
-            foreach (Coin coinSpent in coinsSpent)
-            {
-                //obtain the address relative to this coin (must be improved)
-                HdAddress address = context.UnspentOutputs.First(output => output.ToOutPoint() == coinSpent.Outpoint).Address;
-
-                if (added.Contains(address))
-                    continue;
+            Dictionary<OutPoint,UnspentOutputReference> outpointLookup = context.UnspentOutputs.ToDictionary(o => o.ToOutPoint(), o => o);
```

Comment:
why this?
coinsSpent is a subset of context.UnspentOutputs and prior to some revert that has been done here, it was potentially (and will be when we'll work back on this) the full utxo set that you can spend, that can be tens of thousands of items while coinSpent are just the utxo used in the transaction.

While I agree previous code could be optimized, I don't think this is the correct way:
previous code HdAddress address = context.UnspentOutputs.First(output => output.ToOutPoint() == coinSpent.Outpoint).Address;` was of course sub-obtimal giving the fact it converts UnspentOutputs everytime in the forech, but it's an ienumerable that will stop when find the outpoint it needs, so potentially this change could even be slower.

I agree on the use of a Dictionary but can't we make use of any addressLookup we have now, extending it to be queried against an outpoint too? (I even think address lookup could fit in memory and not be db dependant)

Another way (but I prefers the first) is to create an empty dictionary and populate the dictionary in an iterative way when needed: at each item you check in the dictionary if it exists, if not you continue to check for each UnspentOutputs starting from Dictionary.Count element.
You could even use Linq, skipping Dictionary.Count and then for each element populate the dictionary and return if the output.ToOutPoint() == coinSpent.Outpoint condition is met

Reply:
I'm wondering if we shouldn't have a cache of keys for addresses, since that's the most performance impactful part, it would be paid just by first user of that address, but other calls that involve adding keys to TransactionBuilder would make use of the cache and improve a lot

Reply:
in pseudo code something like this would be nice
```
coinsSpent
   .ForEach(coin => addressLookup.GetAddress(coin))
   .Distinct()
   .ForEach(uniqueAddress => AddKey(uniqueAddress))
```

Reply:
Implemented.

Reply:
adding change to build at commit `b14` on internal mainnet